### ISSUE TITLE
[BUG] Using `qml.state()` with mid-circuit measurements

### ISSUE THREAD

**Issue by isaacdevlugt** [PennyLaneAI/pennylane#4363]
### Expected behavior

When returning `qml.state` from a QNode that has mid-circuit measurements, I expect that the result should be the state / density matrix that one would physically obtain (i.e., a "collapse" occurs).

### Actual behavior

The full state of the system is returned, indicating that no "collapse" occurred. 

### Additional information

Related to https://discuss.pennylane.ai/t/mid-circuit-measurement-in-pennylane/3187

### Source code

```shell
import pennylane as qml

dev = qml.device('default.qubit', wires=2)

@qml.qnode(device=dev)
def bell_state():
    qml.Hadamard(0)
    qml.CNOT(wires=[0,1])
    qml.cond(qml.measure(0), qml.PauliZ)(1)
    return qml.state()

bell_state()
```

### Tracebacks

```shell
tensor([ 0.70710678+0.j,  0.        +0.j,  0.        +0.j,
        -0.70710678+0.j], requires_grad=True)
```

### System information

```shell
Name: PennyLane
Version: 0.31.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Users/isaac/pennylane-stable/lib/python3.9/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning

Platform info:           macOS-13.4.1-x86_64-i386-64bit
Python version:          3.9.14
Numpy version:           1.23.5
Scipy version:           1.10.0
Installed devices:
- default.gaussian (PennyLane-0.31.0)
- default.mixed (PennyLane-0.31.0)
- default.qubit (PennyLane-0.31.0)
- default.qubit.autograd (PennyLane-0.31.0)
- default.qubit.jax (PennyLane-0.31.0)
- default.qubit.tf (PennyLane-0.31.0)
- default.qubit.torch (PennyLane-0.31.0)
- default.qutrit (PennyLane-0.31.0)
- null.qubit (PennyLane-0.31.0)
- lightning.qubit (PennyLane-Lightning-0.31.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by trbromley** [comment]
Now, an error will be raised when returning `qml.state()` with mid-circuit measurements through the deferred measurement principle.

Instead, use `qml.density_matrix(wires=[...])`.
